src/pydvl/influence/array.py

```diff
@@ -0,0 +1,390 @@
+
+    @abstractmethod
+    def from_numpy(self, x: NDArray) -> TensorType:
+        """Overwrite this method for converting a numpy array into a TensorType object"""
+
+
+class TensorAggregator(Generic[TensorType], ABC):
+    """
+    Abstract base class for aggregating tensors. Implementations of this class should define
+    methods for aggregating tensors.
+    """
+
+    @abstractmethod
+    def aggregate_from_nested_generators(
+        self,
+        nested_generators_of_tensors: Generator[
+            Generator[TensorType, None, None], None, None
+        ],
+    ):
+        """
+        Aggregates tensors from a generator of generators.
+
+        Implement this method to specify how tensors, nested in multiple layers of generators,
+        should be combined. Useful for complex data structures where tensors
+        are not directly accessible in a flat list.
+
+        Args:
+            nested_generators_of_tensors: A generator of generators, each yielding TensorType objects."""
+
+    @abstractmethod
+    def aggregate_from_generator(
+        self, tensor_generator: Generator[TensorType, None, None]
+    ):
+        """
+        Aggregates tensors from a single generator.
+
+        Implement this method to define how a sequence of tensors, provided by a generator,
+        should be combined. Suitable for straightforward scenarios where
+        tensors are yielded one by one.
+        """
+
+
+class ListAggregator(TensorAggregator):
+    """
+    An aggregator that collects tensors into list structures.
+    """
+
+    def aggregate_from_nested_generators(
+        self,
+        nested_generators_of_tensors: Generator[
+            Generator[TensorType, None, None], None, None
+        ],
+    ):
+        """
+         Aggregates tensors from a nested generator structure into a list of lists. Each inner generator
+         is converted into a list of tensors, resulting in a nested list structure.
+
+         Args:
+             nested_generators_of_tensors: A generator of generators, where each inner generator yields
+                 TensorType objects.
+
+        Returns:
+            A list of lists, where each inner list contains tensors returned from one of the inner generators.
+        """
+        return [list(tensor_gen) for tensor_gen in nested_generators_of_tensors]
+
+    def aggregate_from_generator(
+        self, tensor_generator: Generator[TensorType, None, None]
+    ):
+        """
+        Aggregates tensors from a single-level generator into a list. This method simply collects
+        each tensor emitted by the generator into a single list.
+
+        Args:
+            tensor_generator: A generator that yields TensorType objects.
+
+        Returns:
+            A list containing all the tensors provided by the tensor_generator.
+        """
+        return [t for t in tensor_generator]
+
+
+class LazyArrayProvider(ABC):
```

Comment:
I don't think this is a "Provider" in the standard sense. There is no global state, nor centralised management of resources nor anything like this. If anything this is a `LazyArrayEvaluator` (please don't use that name 😅). If we don't care about the overloading of the name "Provider" from the design pattern, then maybe `ArrayProvider`, since it is arrays that are provided. But in fact this is a "Realizer" or "Evaluator".